pkg/ottl/ottlfuncs/func_replace_pattern.go

```diff
@@ -47,32 +49,44 @@ func replacePattern[K any](target ottl.GetSetter[K], regexPattern string, replac
 		if err != nil {
 			return nil, err
 		}
-		if fn.IsEmpty() {
-			replacementVal, err = replacement.Get(ctx, tCtx)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			fnVal := fn.Get()
-			replacementExpr, errNew := fnVal.Get(&replacePatternFuncArgs[K]{Input: replacement})
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValStr, ok := replacementValRaw.(string)
-			if !ok {
-				return nil, fmt.Errorf("replacement value is not a string")
-			}
-			replacementVal = replacementValStr
+		replacementVal, err = replacement.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
 		}
 		if originalVal == nil {
 			return nil, nil
 		}
 		if originalValStr, ok := originalVal.(string); ok {
 			if compiledPattern.MatchString(originalValStr) {
+				if !fn.IsEmpty() {
+					match := compiledPattern.FindStringSubmatch(originalValStr)
+					if match != nil {
+						if strings.Contains(replacementVal, "$") {
+							groupNum := strings.ReplaceAll(replacementVal, "$", "")
+							num, _ := strconv.Atoi(groupNum)
+							replacementVal = match[num]
+						}
+					}
```

Comment:
Is there a way to use `Expand` or `ExpandString` here instead?

Reply:
yes, this seems to work
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/5c7a4e94d8c4199b3708d96f6ba3f9258bbfa962